server/src/main/java/io/envoyproxy/controlplane/server/AdsDiscoveryRequestStreamObserver.java

```diff
@@ -0,0 +1,94 @@
+package io.envoyproxy.controlplane.server;
+
+import io.envoyproxy.controlplane.cache.Resources;
+import io.envoyproxy.controlplane.cache.Watch;
+import io.envoyproxy.envoy.api.v2.DiscoveryRequest;
+import io.envoyproxy.envoy.api.v2.DiscoveryResponse;
+import io.grpc.Status;
+import io.grpc.stub.StreamObserver;
+
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.Executor;
+import java.util.function.Supplier;
+
+import static io.envoyproxy.controlplane.server.DiscoveryServer.ANY_TYPE_URL;
+
+public class AdsDiscoveryRequestStreamObserver extends DiscoveryRequestStreamObserver {
+  private final ConcurrentMap<String, Watch> watches;
+  private final ConcurrentMap<String, DiscoveryResponse> latestResponse;
+  private final ConcurrentMap<String, Set<String>> ackedResources;
+
+  AdsDiscoveryRequestStreamObserver(String defaultTypeUrl,
+                                    StreamObserver<DiscoveryResponse> responseObserver,
+                                    long streamId,
+                                    Executor executor,
+                                    DiscoveryServer discoveryServer) {
+    super(defaultTypeUrl, responseObserver, streamId, executor, discoveryServer);
+    this.watches = new ConcurrentHashMap<>(Resources.TYPE_URLS.size());
+    this.latestResponse = new ConcurrentHashMap<>(Resources.TYPE_URLS.size());
+    this.ackedResources = new ConcurrentHashMap<>(Resources.TYPE_URLS.size());
+  }
+
+  @Override
+  public void onNext(DiscoveryRequest request) {
+    String requestTypeUrl = request.getTypeUrl();
+
+    if (defaultTypeUrl.equals(ANY_TYPE_URL)) {
+      if (requestTypeUrl.isEmpty()) {
+        responseObserver.onError(
+            Status.UNKNOWN
+                .withDescription(String.format("[%d] type URL is required for ADS", streamId))
+                .asRuntimeException());
+
+        return;
+      }
+    } else if (requestTypeUrl.isEmpty()) {
+      requestTypeUrl = defaultTypeUrl;
+    }
```

Comment:
I think you can simplify it. For the `defaultTypeUrl` you always pass `ANY_TYPE_URL`. You can delete the `defaultTypeUrl` from the constructor, pass the `ANY_TYPE_URL` to `DiscoveryRequestStreamObserver`.
Then you can delete `if (defaultTypeUrl.equals(ANY_TYPE_URL)) {` if and also `} else if (requestTypeUrl.isEmpty()) {` branch

Reply:
that is correct, good catch 👍